core/modules/replica/StatusRequest.cc

```diff
@@ -0,0 +1,206 @@
+#include <boost/bind.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
+
+// Qserv headers
+#include "lsst/log/Log.h"
+#include "replica/ProtocolBuffer.h"
+#include "replica/ServiceProvider.h"
+
+// This macro to appear witin each block which requires thread safety
+#define LOCK_GUARD std::lock_guard<std::mutex> lock(_mtx)
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.StatusRequest");
+
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+StatusRequestBaseM::StatusRequestBaseM(ServiceProvider::pointer const& serviceProvider,
+                                       boost::asio::io_service& io_service,
+                                       char const*              requestTypeName,
+                                       std::string const&       worker,
+                                       std::string const&       targetRequestId,
+                                       proto::ReplicationReplicaRequestType requestType,
+                                       bool                     keepTracking,
+                                       std::shared_ptr<Messenger> const& messenger)
+    :   RequestMessenger(serviceProvider,
+                         io_service,
+                         requestTypeName,
+                         worker,
+                         0,    /* priority */
+                         keepTracking,
+                         false /* allowDuplicate */,
+                         messenger),
+        _targetRequestId(targetRequestId),
+        _requestType(requestType) {
+}
+
+void StatusRequestBaseM::startImpl() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "startImpl");
+
+    // Serialize the Request message header and the request itself into
+    // the network buffer.
+
+    _bufferPtr->resize();
+
+    proto::ReplicationRequestHeader hdr;
+    hdr.set_id(id());
+    hdr.set_type(proto::ReplicationRequestHeader::REQUEST);
+    hdr.set_management_type(proto::ReplicationManagementRequestType::REQUEST_STATUS);
+
+    _bufferPtr->serialize(hdr);
+
+    proto::ReplicationRequestStatus message;
+    message.set_id(_targetRequestId);
+    message.set_type(_requestType);
+
+    _bufferPtr->serialize(message);
+
+    send();
+}
+
+void StatusRequestBaseM::wait() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "wait");
+
+    // Allways need to set the interval before launching the timer.
+
+    _timer.expires_from_now(boost::posix_time::seconds(_timerIvalSec));
+    _timer.async_wait(
+        boost::bind(
+            &StatusRequestBaseM::awaken,
+            shared_from_base<StatusRequestBaseM>(),
+            boost::asio::placeholders::error
+        )
+    );
+}
+
+void StatusRequestBaseM::awaken(boost::system::error_code const& ec) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "awaken");
+
+    if (isAborted(ec)) { return; }
+
+    // Also ignore this event if the request expired
+    if (_state== State::FINISHED) { return; }
+
+    LOCK_GUARD;
```

Comment:
couldn't the state change after the test and before taking this lock?

Reply:
For the record. This was a very good point, and it was a potential problem in the code. It has been fixed in this and all similar files. Thanks!